nemo/collections/asr/parts/nmse_clustering.py

```diff
@@ -0,0 +1,221 @@
+    """
+    eps = 1e-10
+    eig_ratio_list = []
+    if fixed_thres:
+        p_neighbors_list = [int(mat.shape[0] * fixed_thres)]
+        max_N = p_neighbors_list[0]
+    else:
+        max_N = int(mat.shape[0] * max_rp_threshold)
+        if sparse_search:
+            N = min(max_N, search_p_volume)
+            p_neighbors_list = list(np.linspace(1, max_N, N, endpoint=True).astype(int))
+        else:
+            p_neighbors_list = list(range(1, max_N))
+
+    est_spk_n_dict = {}
+    for p_neighbors in p_neighbors_list:
+        X_conn_from_dist = get_X_conn_from_dist(mat, p_neighbors)
+        est_num_of_spk, lambdas, lambda_gap_list = estimate_num_of_spkrs(X_conn_from_dist, SPK_MAX)
+        est_spk_n_dict[p_neighbors] = (est_num_of_spk, lambdas)
+        arg_sorted_idx = np.argsort(lambda_gap_list[:SPK_MAX])[::-1]
+        max_key = arg_sorted_idx[0]
+        max_eig_gap = lambda_gap_list[max_key] / (max(lambdas) + eps)
+        eig_ratio_value = (p_neighbors / mat.shape[0]) / (max_eig_gap + eps)
+        eig_ratio_list.append(eig_ratio_value)
+
+    index_nn = np.argmin(eig_ratio_list)
+    rp_p_neighbors = p_neighbors_list[index_nn]
+    X_conn_from_dist = get_X_conn_from_dist(mat, rp_p_neighbors)
+    if not isFullyConnected(X_conn_from_dist):
+        X_conn_from_dist, rp_p_neighbors = gc_thres_min_gc(mat, max_N, p_neighbors_list)
+
+    return (
+        X_conn_from_dist,
+        float(rp_p_neighbors / mat.shape[0]),
+        est_spk_n_dict[rp_p_neighbors][0],
+        est_spk_n_dict[rp_p_neighbors][1],
+        rp_p_neighbors,
+    )
+
+
+def get_eigen_matrix(emb):
+    """
+    input
+    emb: embedding
+    
+    output
+    similarity matrix
+    """
+    sim_d = cosine_similarity(emb)
+    scaler.fit(sim_d)
+    sim_d = scaler.transform(sim_d)
+
+    return sim_d
+
+
+def COSclustering(key, emb, oracle_num_speakers=None, max_num_speaker=8):
+    """
+    input:
+    key (str): speaker uniq name
+    emb (np array): speaker embedding
+    oracle_num_speaker (int or None): oracle number of speakers if known else None
+    max_num_speakers (int): maximum number of clusters to consider for each session
+    output:
+    Y (List[int]): speaker labels
+    """
+    est_num_spks_out_list = []
+    mat = get_eigen_matrix(emb)
+
+    if oracle_num_speakers:
+        max_num_speaker = oracle_num_speakers
+
+    X_conn_spkcount, rp_thres_spkcount, est_num_of_spk, lambdas, p_neigh = NMEanalysis(mat, max_num_speaker)
+    X_conn_from_dist = get_X_conn_from_dist(mat, p_neigh)
+
+    if oracle_num_speakers:
+        est_num_of_spk = oracle_num_speakers
+
+    est_num_spks_out_list.append([key, str(est_num_of_spk)])
+
+    # Perform spectral clustering
+    spectral_model = sklearn_SpectralClustering(
+        affinity='precomputed',
+        eigen_solver='amg',
+        random_state=0,
+        n_jobs=-1,
+        n_clusters=est_num_of_spk,
+        eigen_tol=1e-10,
+    )
+
+    Y = spectral_model.fit_predict(X_conn_from_dist)
```

Comment:
Remove all these java/c type function names and change to snake case. Improve docs throughout the entire file in Google docstring format.

Reply:
As said, this file contains code from sklearn repository as mentioned on top of the page. so left them as is. As its easy for copyright purposes. These functions are internal and wont be called or changed by any user specifically